### PR TITLE
decouple aenea server from platform dependent rpcs

### PR DESCRIPTION
**Pull request by mzizzi**
Sparked by discussion in #105 we could support multiple implementations for a single platform and/or share more code between server implementations.  This PR is a step towards a more modular and reusable server.  I don't expect it to be 100% ready for merge but more of a starting point for discussion as we continue to make #105 ready for prime time.

The good:
  * Provides an interface that provides stubs for all of Aenea's core RPC methods.  This is a "grab and go" way other developers to implement Aenea for other platforms.  Developers for non-linux platforms should worry less about making a compatible jsonrpc server and more about implementing the RPCs themselves. (Not that this is a major problem now but it's nice.)
  * Tools to automatically load yapsy plugins for the aenea server.  Again, we should aim to solve these common problems only once.
  * Tools for automatically configuring basic console/file logging for the aenea server.
  * A _much_ more testable core server.  We can easily inject mocks at several different layers for testing.
  * We can generate "one size fits all" docs from the RPC stubs to help explain the project.

The bad:
  * Although I believe this code will lead to less code overall in the long run it is certainly _more_ code right now.  Less code is typically better.  More code == more liability.
  * More configuration options could cause pain.

### CODE REVIEW

**Comment by calmofthestorm** [comment]
I'll need to think about this. My instinct is that this is over-engineering for the problem at hand -- this is just a 500 line Python script (before your other PR, which makes it much shorter) that's basically just a thin wrapper around xdotool/libxdo and jsonrpclib. Integration testing is going to be entirely platform specific anyway, and unit testing seems pointless to me on a script this simple. It's not sloc that scares me about this, it's the added complexity.

What logic (current or anticipated) would be shared between server implementations? Off the top of my head:

- yapsy plugin loading
- logging
- jsonrpc code (but honestly jsonrpclib handles nearly all of this for users)
- authentication (not currently implemented, but I think it's something I should probably do at least a basic version of)

None of these are big, and they could easily be wrapped in their own module instead of having an interface tightly coupling them together.

On the other hand, even if we do merge this there's no reason a user could not implement their own server to the spec writing it all themselves, so it's not clear to me there's a downside.

Let me read it over again and think about it.

**Comment by mzizzi** [comment]
I won't take offense to tossing this one.  It does feel a bit over engineered.  It took me writing it to realize that, however.

> What logic (current or anticipated) would be shared between server implementations? Off the top of my head

Everything you mentioned plus:
  * Encryption
  * Possibly some of the higher level and yet to be decided on key translations from #65

File: server/core.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+    over the network.
+    """
+    def __init__(self, rpc_impl, server, plugins=tuple(), logger=None):
+        """
+        :param rpc_impl: Object that implements all AbstractAeneaPlatformRpc
+         methods.  This is where the platform specific magic happens to gather
+         context and emulate input.
+        :param SimpleJSONRPCServer server: rpcs from <rpc_impl> will be attached
+          to this server.
+        :param plugins: yapsy plugin objects.  Plugin objects should
```
Comment (calmofthestorm) [inline server/core.py]: What should they do?

**Comment by mzizzi** [comment]
Good catch, thanks.

File: server/core.py (hunk 1 above)
Comment (mzizzi) [inline server/core.py]: Good catch, thanks.

**Comment by calmofthestorm** [comment]
While I still think it's probably overkill, and am uneasy about the wisdom of sharing code across platforms, I think merging this is reasonable for breaking up the server script into logical chunks. The main drawback to this in my opinion would be that I don't think that sharing code across platforms is necessarily realistic, but fundamentally merging this doesn't force a Windows or OS X server to use it, merely makes it available.

I have one further concern before merging this: How are you supposed to run it so that server_x11.py can import server.core? I solved this for testing by symlinking server into the linux_x11 dir and running from there but my guess is that this is not your intent?

**Comment by mzizzi** [comment]
> While I still think it's probably overkill, and am uneasy about the wisdom of sharing code across platforms, I think merging this is reasonable for breaking up the server script into logical chunks.

Maybe this is overkill.  I like that there is a clear separation of responsibilities for input and rpcs.  I dislike that it requires more mental overhead to understand how the server works.  The question we should be asking is this, "is code sharing between platforms and the separation of responsibilities worth the added complexity?" 

> The main drawback to this in my opinion would be that I don't think that sharing code across platforms is necessarily realistic

I'm not sure I follow why sharing server code cross-platform be a bad thing.  You're correct in your assertion that most of the meaningful server code is in input emulation but I don't see why that means that the project should support three different ways of bootstrapping the boring parts of a server.  If this were a larger project we would be much better off supporting a single core server than three.  

> How are you supposed to run it so that server_x11.py can import server.core?
I didn't realize I broke this.  My run config for the server adds the root of the project to `PYTHONPATH` so I didn't see the issue.  You can solve with without a symlink like so:

```bash
PYTHONPATH=/path/to/aenea python server/linux_x11/server_x11.py
```

I'll see what I can do to make this easier.

**Comment by calmofthestorm** [comment]
> I didn't realize I broke this. My run config for the server adds the root of the project to PYTHONPATH so I didn't see the issue. You can solve with without a symlink like so:
> 
> `PYTHONPATH=/path/to/aenea python server/linux_x11/server_x11.py`
> 
> I'll see what I can do to make this easier.
If you just add this to the README where we talk about running that server that should be fine.

**Comment by calmofthestorm** [comment]
In case I wasn't clear, I think we should merge this. I may have reservations about some of the things we've discussed, but I feel that breaking the code up like this is a win on its own because, as you say, it breaks up responsibilities.

Only thing I'm waiting on is fixing the issue I had with PYTHONPATH. That could be as simple as updating the readme to give the invocation, or some other solution.

**Comment by mzizzi** [comment]
I'd rather keep backwards compatibility for those who already use the server_x11.py.  34ca271 adds a quick hack to dynamically add the project root to the path at run time.

**Comment by mzizzi** [comment]
I snuck in one last quick change prior to merging.  Looking back at the core code it seemed a bit overkill to introduce a "ServerFactory" that would only ever have a single method hanging off it it.  f708103 moves the ServerFactory into a classmethod.

**Comment by calmofthestorm** [comment]
Perfect, thanks!

**Pull request merged**